core/src/main/java/org/apache/iceberg/io/ReadBuilder.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.mapping.NameMapping;
+
+/**
+ * File formats should implement this interface to provide a builder for reading data files. {@link
+ * ReadBuilder} reads the data files with the specified parameters. The returned objects are defined
+ * by the {@link ObjectModel} which is used to read the data.
+ *
+ * <p>This interface is directly exposed for the users to parameterize readers.
+ *
+ * @param <B> type returned by builder API to allow chained calls
+ */
+public interface ReadBuilder<B extends ReadBuilder<B>> {
```

Comment:
Do we need to add support for custom types here?  This is necessary for metadata/stats/etc.

Reply:
I guess you are suggesting to change the `<D> CloseableIterable<D> build()` method to return some other object, like `<D> FileReader<D> build();`, and the `FileReader` could have `data`/`metadata`/`stats` attributes, like:
```
public interface FileReader<D> {
  CloseableIterable<D> data();
  Map<String, String> metadata();
  Stat statistics();
}
```

The issue is that the build() method is a public API, and widely used. We could deprecate it, and create a new return type, but I would suggest that do it in a different PR.

Did I understand your suggestion correctly?
WDYT?

Thanks for the review!

Reply:
In a follow-up PR we could do something like this:
```
public interface ReadBuilder<B extends ReadBuilder<B>> {
[..]
  /** Builds the reader. */
  <D> FileReadIterator<D> build();

  public interface FileReadIterator<D> extends CloseableIterable<D> {
    Map<String, String> metadata();
  }
}
```